src/app/libp2p_helper/src/libp2p_helper/main.go

```diff
@@ -1138,57 +1148,139 @@ func (a *app) checkBandwidth(id peer.ID) {
+		app.P2p.Logger.Debugf("couldn't register rate-out bandwidth gauge for id", id, "perhaps we've already done so", err.Error())
 		return
 	}
 
 	for {
-		stats := a.P2p.BandwidthCounter.GetBandwidthForPeer(id)
+		stats := app.P2p.BandwidthCounter.GetBandwidthForPeer(id)
 		totalIn.Set(float64(stats.TotalIn))
 		totalOut.Set(float64(stats.TotalOut))
 		rateIn.Set(stats.RateIn)
 		rateOut.Set(stats.RateOut)
 
-		time.Sleep(metricsRefreshTime)
+		time.Sleep(app.MetricsRefreshTime)
+	}
+}
+
+func (app *app) checkPeerCount(id peer.ID) {
+	peerCountGauge := prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: fmt.Sprintf("peer_count_%s", id),
+		Help: "The total number of peers connected by the peer",
+	})
+
+	err := prometheus.Register(peerCountGauge)
+	if err != nil {
+		app.P2p.Logger.Debugf("couldn't register peer count for id", id, "perhaps we've already done so", err.Error())
+		return
+	}
+
+	for {
+		peerIDs := app.P2p.Host.Network().Peers()
+		peerCountGauge.Set(float64(len(peerIDs)))
+		time.Sleep(app.MetricsRefreshTime)
+	}
+}
+
+func (app *app) checkMessageExchanged(id peer.ID) {
+	msgGauge := prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: fmt.Sprintf("message_exchanged_%s", id),
+		Help: "The number of different peers that the peer connects with",
+	})
+
+	err := prometheus.Register(msgGauge)
+	if err != nil {
+		app.P2p.Logger.Debugf("couldn't register message exchanged of peers for id", id, "perhaps we've already done so", err.Error())
+		return
+	}
+
+	for {
+		currentCount := app.P2p.ConnectionManager.GetInfo().ConnCount
+		msgGauge.Set(float64(currentCount))
+		time.Sleep(app.MetricsRefreshTime)
+	}
+}
+
+func (app *app) checkMessageStats(id peer.ID) {
+	msgMaxGauge := prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: fmt.Sprintf("message_max_stats_%s", id),
+		Help: "The max size of network message received by the peer",
+	})
+
+	msgAvgGauge := prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: fmt.Sprintf("message_avg_stats_%s", id),
+		Help: "The average size of network message received by the peer",
+	})
+
+	msgMinGauge := prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: fmt.Sprintf("message_min_stats_%s", id),
+		Help: "The min size of network message received by the peer",
+	})
+
+	err := prometheus.Register(msgMaxGauge)
+	if err != nil {
+		app.P2p.Logger.Debugf("couldn't register message max stats for id", id, "perhaps we've already done so", err.Error())
+		return
+	}
+
+	err = prometheus.Register(msgAvgGauge)
+	if err != nil {
+		app.P2p.Logger.Debugf("couldn't register message average stats for id", id, "perhaps we've already done so", err.Error())
+		return
+	}
+
+	err = prometheus.Register(msgMinGauge)
+	if err != nil {
+		app.P2p.Logger.Debugf("couldn't register message min stats for id", id, "perhaps we've already done so", err.Error())
+		return
+	}
+
+	for {
+		msgStats := app.P2p.MsgStats
+		msgMinGauge.Set(float64(msgStats.GetMin()))
```

Comment:
These are always called in succession, and each performs a `RLock`. Can we instead have one method `GetStats` which only does a single `RLock` and returns all 3 values in a struct?

Reply:
done!